server/server_test.go

```diff
@@ -29,38 +32,476 @@ func performRequest(r http.Handler, method, path string, body io.Reader) (*httpt
+				continue
+			}
+			expPlayer := model.Player{
+				ID:   model.PlayerID(r.PlayerID),
+				Name: r.DispName,
+			}
+			var player model.Player
+			readBody(t, w.Body, &player)
+			assert.NoError(t, err)
+			assert.Equal(t, expPlayer, player)
+		}
+	}
+}
+func TestGinPostCreateGame(t *testing.T) {
 	testCases := []struct {
 		msg     string
-		data    network.CreatePlayerModel
-		expRes  model.Player
+		pIDs    []string
 		expCode int
-	}{
-		/*
-			TODO add test cases once the server is more testable. We need to rewrite it to inject persistence in so we can mock it here
-			(or at least set it to memory instead of mongo by default)
-		*/
+		expErr  string
+	}{{
+		msg:     `two player game`,
+		pIDs:    []string{`p1`, `p2`},
+		expCode: http.StatusOK,
+		expErr:  ``,
+	}, {
+		msg:     `three player game`,
+		pIDs:    []string{`p1`, `p2`, `p3`},
+		expCode: http.StatusOK,
+		expErr:  ``,
+	}, {
+		msg:     `four player game`,
+		pIDs:    []string{`p1`, `p2`, `p3`, `p4`},
+		expCode: http.StatusOK,
+		expErr:  ``,
+	}, {
+		msg:     `one player game is an error`,
+		pIDs:    []string{`p1`},
+		expCode: http.StatusBadRequest,
+		expErr:  `Invalid num players: 1`,
+	}, {
+		msg:     `five player game`,
+		pIDs:    []string{`p1`, `p2`, `p3`, `p4`, `p5`},
+		expCode: http.StatusBadRequest,
+		expErr:  `Invalid num players: 5`,
+	}, {
+		msg:     `zero player game`,
+		pIDs:    []string{},
+		expCode: http.StatusBadRequest,
+		expErr:  `Invalid num players: 0`,
+	}, {
+		msg:     `missing player id`,
+		pIDs:    []string{`p1`, `p2`, ``, `p4`},
+		expCode: http.StatusBadRequest,
+		expErr:  `Invalid player ID at index 2`,
+	}, {
+		msg:     `invalid player id`,
+		pIDs:    []string{`p1`, `p2`, `#`, `p4`},
+		expCode: http.StatusInternalServerError,
+		expErr:  `createGame error: player not found`,
+	}, {
+		msg:     `create a game with nonexistent players`,
+		pIDs:    []string{`p1`, `p6`},
+		expCode: http.StatusInternalServerError,
+		expErr:  `createGame error: player not found`,
+	}, {
+		msg:     `bad request body - equivalent to having no player IDs`,
+		pIDs:    []string{},
+		expCode: http.StatusBadRequest,
+		expErr:  `Invalid num players: 0`,
+	}}
+	cs, router := newServerAndRouter()
+	// seed the db with players
+	seedPlayers(t, cs.db, 5)
+	for _, tc := range testCases {
+		g := model.Game{}
+		g.Players = make([]model.Player, len(tc.pIDs))
+		for i, id := range tc.pIDs {
+			g.Players[i] = model.Player{ID: model.PlayerID(id)}
+		}
+		// make the request
+		body := prepareBody(t, g)
+		w, err := performRequest(router, `POST`, `/create/game`, body)
+		require.NoError(t, err)
+		// verify
+		require.Equal(t, tc.expCode, w.Code)
+		if tc.expCode != http.StatusOK {
+			errMsg := readError(t, w)
+			assert.Equal(t, tc.expErr, errMsg)
+			continue
+		}
+		var game model.Game
+		readBody(t, w.Body, &game)
+		// verify the players are in the game
+		for _, p := range g.Players {
+			_, ok := game.PlayerColors[p.ID]
+			assert.True(t, ok)
```

Comment:
This _doesn't_ need to be true, but it is currently.. 🤷

Reply:
🤔shouldn't the players from the `network.CreateGameRequest` be in the game?

Reply:
IRL we can start a game and _not_ have colors assigned to each player. we could wait to assign colors. 🤷

Reply:
🤦 duh. I'll check the players instead